### PR TITLE
feat: modify auth token url based on instance param

### PR DESCRIPTION
**Pull request by cmars**
If an instance parameter is provided in the redirect, use it to modify
the URL from where the oauth token is obtained.

The instance provided is the Snyk region domain, minus the api. host
prefix.

Stacked on #258

### CODE REVIEW

File: pkg/auth/oauth2authenticator.go [hunk 1]
```diff
@@ -402,6 +423,16 @@ func (o *oAuth2Authenticator) authenticateWithAuthorizationCode() error {
 	return err
 }
 
+func redirectAuthHost(instance string) string {
+	return fmt.Sprintf("api.%s", instance)
+}
+
+var redirectAuthHostRE = regexp.MustCompile(`^api\.(.+)\.snyk\.io$`)
```
Comment (DarrellMozingo) [inline pkg/auth/oauth2authenticator.go:430]: snykgov too!

File: pkg/app/app.go [hunk 2]
```diff
@@ -91,6 +101,61 @@ func defaultFuncApiUrl(logger *zerolog.Logger) configuration.DefaultValueFunctio
...
+		// NOTE: The original jws package models audience with a string, not a
+		// []string. This fails to parse Snyk JWTs.
+		Aud []string `json:"aud"`
+	}{}
+	err = json.NewDecoder(bytes.NewBuffer(decoded)).Decode(&c)
+	if err != nil {
+		return "", err
+	}
+	if len(c.Aud) > 0 {
+		return c.Aud[0], nil
```
Comment (ShawkyZ) [inline pkg/app/app.go:154]: could we by any change have multiple entries in the `aud` claim ?

File: pkg/app/app.go [hunk 3]
```diff
@@ -91,6 +101,61 @@ func defaultFuncApiUrl(logger *zerolog.Logger) configuration.DefaultValueFunctio
...
+		return "", errors.New("jws: invalid token received")
+	}
+	decoded, err := base64.RawURLEncoding.DecodeString(s[1])
+	if err != nil {
+		return "", err
+	}
+	c := struct {
+		// NOTE: The original jws package models audience with a string, not a
+		// []string. This fails to parse Snyk JWTs.
+		Aud []string `json:"aud"`
```
Comment (ShawkyZ) [inline pkg/app/app.go:147]: it doesn't feel correct to use the `aud` claim for fetching tokens.
We are using aud claim to determine where to fetch the access token from. 
Afaik aud claim should only be used to identify the audience/recipients for this token.
https://datatracker.ietf.org/doc/html/rfc7519#section-4.1.3

File: pkg/app/app.go (hunk 2 above)
Comment (thisislawatts) [reply pkg/app/app.go:154]: From the spec:
>  In the general case, the "aud" value is an array of case-
   sensitive strings, each containing a StringOrURI value.  In the
   special case when the JWT has one audience, the "aud" value MAY be a
   single case-sensitive string containing a StringOrURI value.  The
   interpretation of audience values is generally application specific.
   Use of this claim is OPTIONAL.
 https://datatracker.ietf.org/doc/html/rfc7519#section-4.1.3
  
 It is almost expected for there to be multiple values available. We should likely search the list and extract the first one which meets our criteria.

Comment (cmars) [reply pkg/app/app.go:154]: Guidance from the IdP team is to use the first value. I can provide more context offline.

**Review by thisislawatts — APPROVED** [review]

**Review by PeterSchafer — APPROVED** [review]

**Pull request merged**